packages/core/src/server/index.ts

```diff
@@ -208,31 +208,30 @@ export class DevServer implements ImplDevServer {
         const onError = async (error: { code: string }) => {
           if (error.code === 'EADDRINUSE') {
             clearScreen();
+            logger.warn(`Port ${devPort} is in use, trying another one...`);
             resolve(false);
           } else {
-            resolve(true);
+            reject(true);
           }
         };
-        if (strictPort) {
+        httpServer.on('error', onError);
+        httpServer.on('listening', () => {
           httpServer.removeListener('error', onError);
-          reject(new Error(`Port ${devPort} is already in use`));
-        } else {
-          logger.warn(`Port ${devPort} is in use, trying another one...`);
-          httpServer.on('error', onError);
-          httpServer.on('listening', () => {
-            httpServer.close();
-            resolve(true);
-          });
-          httpServer.listen(portToCheck, '::1');
-        }
+          httpServer.close();
+          resolve(true);
+        });
+        httpServer.listen(portToCheck, 'localhost');
       });
     };
-
-    let isPortAvailableResult = await isPortAvailable(devPort);
-    while (isPortAvailableResult === false) {
-      userConfig.server.hmr = { port: ++hmrPort };
-      userConfig.server.port = ++devPort;
-      isPortAvailableResult = await isPortAvailable(devPort);
+    if (strictPort) {
+      throw new Error(`Port ${devPort} is already in use`);
+    } else {
+      let isPortAvailableResult = await isPortAvailable(devPort);
+      while (isPortAvailableResult === false) {
+        userConfig.server.hmr = { port: ++hmrPort };
```

Comment:
If the server user is not configured, will userConfig.server is undefined appear?